File: java/plugins/quickfiximpl/src/main/java/com/applicationsx/quickfiximpl/QuickFixClient.java
package com.applicationsx.quickfiximpl;

import quickfix.*;
import quickfix.field.*;
import quickfix.fix44.ExecutionReport;
import quickfix.fix44.NewOrderSingle;

import java.time.LocalDateTime;

public class QuickFixClient extends MessageCracker implements Application {
    private SessionID sessionId;

    @Override
    public void onCreate(SessionID sessionID) {
        this.sessionId = sessionID;
        System.out.println("Session created: " + sessionID);
    }

    @Override
    public void onLogon(SessionID sessionID) {
        System.out.println("Logged on: " + sessionID);
    }

    @Override
    public void onLogout(SessionID sessionID) {
        System.out.println("Logged out: " + sessionID);
    }

    @Override
    public void toAdmin(Message message, SessionID sessionID) {
        // Add admin messages if needed
    }

    @Override
    public void fromAdmin(Message message, SessionID sessionID) {
        // Handle admin messages
    }

    @Override
    public void toApp(Message message, SessionID sessionID) {
        // Add application messages if needed
    }

    @Override
    public void fromApp(Message message, SessionID sessionID) {
        try {
            crack(message, sessionID);
        } catch (UnsupportedMessageType | FieldNotFound | IncorrectTagValue e) {
            throw new RuntimeException(e);
        }
    }

    public void onMessage(ExecutionReport message, SessionID sessionID) {
        System.out.println("Received execution report: " + message);
    }

    public void sendNewOrder(String symbol, char side, double quantity, double price) throws SessionNotFound {
        NewOrderSingle order = new NewOrderSingle();
        order.set(new ClOrdID("ORDER-" + System.currentTimeMillis()));
        order.set(new Symbol(symbol));
        order.set(new Side(side));
        order.set(new OrdType(OrdType.LIMIT));
        order.set(new OrderQty(quantity));
        order.set(new Price(price));
        order.set(new TimeInForce(TimeInForce.DAY));
        order.set(new TransactTime(LocalDateTime.now()));

        Session.sendToTarget(order, sessionId);
    }
} 

File: cpp/src/quickfiximpl/QuickFixClient.cpp
#include "QuickFixClient.h"
#include <chrono>
#include <sstream>

QuickFixClient::QuickFixClient() : session(nullptr) {}

void QuickFixClient::onCreate(const FIX::SessionID& sessionID) {
    session = FIX::Session::lookupSession(sessionID);
    std::cout << "Client session created: " << sessionID << std::endl;
}

void QuickFixClient::onLogon(const FIX::SessionID& sessionID) {
    std::cout << "Client logged on: " << sessionID << std::endl;
}

void QuickFixClient::onLogout(const FIX::SessionID& sessionID) {
    std::cout << "Client logged out: " << sessionID << std::endl;
}

void QuickFixClient::toAdmin(FIX::Message& message, const FIX::SessionID& sessionID) {
    // Add admin-level message handling if needed
}

void QuickFixClient::fromAdmin(const FIX::Message& message, const FIX::SessionID& sessionID)
    throw(FIX::FieldNotFound, FIX::IncorrectDataFormat, FIX::IncorrectTagValue, FIX::RejectLogon) {
    // Handle admin-level messages
}

void QuickFixClient::toApp(FIX::Message& message, const FIX::SessionID& sessionID)
    throw(FIX::DoNotSend) {
    // Add application-level message handling if needed
}

void QuickFixClient::fromApp(const FIX::Message& message, const FIX::SessionID& sessionID)
    throw(FIX::FieldNotFound, FIX::IncorrectDataFormat, FIX::IncorrectTagValue, FIX::UnsupportedMessageType) {
    crack(message, sessionID);
}

void QuickFixClient::onMessage(const FIX44::ExecutionReport& message, const FIX::SessionID& sessionID)
    throw(FIX::FieldNotFound) {
    FIX::ClOrdID clOrdID;
    message.get(clOrdID);
    std::cout << "Received execution report: " << clOrdID.getValue() << std::endl;
}

void QuickFixClient::sendNewOrder(const std::string& symbol, char side, double quantity, double price) {
    FIX44::NewOrderSingle order;
    
    // Generate unique order ID
    auto now = std::chrono::system_clock::now();
    auto now_ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch());
    std::stringstream ss;
    ss << "ORDER" << now_ms.count();
    
    order.set(FIX::ClOrdID(ss.str()));
    order.set(FIX::HandlInst('1'));
    order.set(FIX::Symbol(symbol));
    order.set(FIX::Side(side));
    order.set(FIX::TransactTime());
    order.set(FIX::OrdType(FIX::OrdType_LIMIT));
    order.set(FIX::TimeInForce(FIX::TimeInForce_DAY));
    order.set(FIX::OrderQty(quantity));
    order.set(FIX::Price(price));

    if (session) {
        FIX::Session::send(order);
    }
} 